# Data Cleaning with Pandas

In [2]:
import pandas as pd
import numpy as np

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

In [1]:
!pwd
!cd data
!pwd

/Users/dakaspar/flatiron/pandas-3-data-cleaning-seattle-ds-career-040119
/Users/dakaspar/flatiron/pandas-3-data-cleaning-seattle-ds-career-040119


### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [3]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [8]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [4]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [5]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [6]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [4]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [5]:
bldg_df.head().T   # I ran this out of order after I had already dropped a bunch of columns

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [7]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [8]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [8]:
#sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])   #throws error, already dealt with it later

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [9]:
#sales_df['Major'] = pd.to_numeric(sales_df['Major'])   #throws error, already dealt with it later

ValueError: Unable to parse string "      " at position 935519

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [16]:
# The single question mark means "show me the docstring"
#pd.to_numeric?

Signature: pd.to_numeric(arg, errors='raise', downcast=None)
Docstring:
Convert argument to a numeric type.

Parameters
----------
arg : list, tuple, 1-d array, or Series
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
downcast : {'integer', 'signed', 'unsigned', 'float'} , default None
    If not None, and if the data has been successfully cast to a
    numerical dtype (or if the data was numeric to begin with),
    downcast that resulting data to the smallest numerical dtype
    possible according to the following rules:

    - 'integer' or 'signed': smallest signed int dtype (min.: np.int8)
    - 'unsigned': smallest unsigned int dtype (min.: np.uint8)
    - 'float': smallest float dtype (min.: np.float32)

    As this behaviour is separate from the core conversion to
    numeric values, any er

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [9]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [10]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [11]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [12]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [13]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [14]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [15]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [17]:
#sales_data.loc[(sales_data['SqFtTotLiving'] == 780) & (sales_data['ZipCode'] == 98103)]

ExciseTaxNbr     Major   Minor DocumentDate  SalePrice  \
354508        2739349  336290.0   365.0   06/17/2015     450000   
354509        2332157  336290.0   365.0   02/01/2008          0   
354510        2187702  336290.0   365.0   02/21/2006          0   
354511        2330224  336290.0   365.0   01/24/2008     482500   
354512        1236189  336290.0   365.0   03/13/1992     128000   
354513        2409079  336290.0   365.0   07/16/2009          0   
1086217       1350395  379700.0  1481.0   12/14/1993          0   
1086218       2881988  379700.0  1481.0   08/03/2017          0   
1086219       2915482  379700.0  1481.0   02/05/2018     590000   
1086220       2874899  379700.0  1481.0   07/02/2017          0   
1086221       2205679  379700.0  1481.0   04/08/2006          0   
1086222       2881987  379700.0  1481.0   08/03/2017     920000   
1086223       2904774  379700.0  1481.0   11/30/2017     600000   
1086224       2846576  379700.0  1481.0   01/24/2017     875000   
1426481       1739606  812970.0   750.0   03/01/2000     235000   

           RecordingNbr Volume Page PlatNbr PlatType   ...   SaleInstrument  \
354508   20150626001832                                ...                3   
354509   20080211001755                                ...                3   
354510   20060221002503                                ...               15   
354511   20080128001148                                ...                3   
354512   199203201310      010  043  336290        P   ...                2   
354513   20090922001177                                ...               15   
1086217  199312283773      010  089  379700        P   ...               15   
1086218                                                ...               15   
1086219                                                ...                3   
1086220  20170703000480                                ...               15   
1086221  20060510000308                                ...               15   
1086222                                                ...                3   
1086223  20171207000996                                ...                3   
1086224  20170130001087                                ...                3   
1426481  20000303000728                                ...                3   

        AFForestLand AFCurrentUseLand AFNonProfitUse  AFHistoricProperty  \
354508             N                N              N                   N   
354509             N                N              N                   N   
354510             N                N              N                   N   
354511             N                N              N                   N   
354512             N                N              N                   N   
354513             N                N              N                   N   
1086217            N                N              N                   N   
1086218            N                N              N                   N   
1086219            N                N              N                   N   
1086220            N                N              N                   N   
1086221            N                N              N                   N   
1086222            N                N              N                   N   
1086223            N                N              N                   N   
1086224                                                                    
1426481            N                N              N                   N   

         SaleReason  PropertyClass  SaleWarning SqFtTotLiving ZipCode  
354508            1              8                        780   98103  
354509           14              8     31 51 52           780   98103  
354510           10              8     18 31 51           780   98103  
354511            1              8                        780   98103  
354512            1              8                        780   98103  
354513           14   

In [16]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [18]:
sd = sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [55]:
sd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1322148 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1322148 non-null float64
Minor            1322148 non-null float64
DocumentDate     1322148 non-null object
SalePrice        1322148 non-null int64
SqFtTotLiving    1322148 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 70.6+ MB


In [56]:
sd.describe()

Major         Minor     SalePrice  SqFtTotLiving
count  1.322148e+06  1.322148e+06  1.322148e+06   1.322148e+06
mean   4.477971e+05  1.588399e+03  2.826850e+05   2.088334e+03
std    2.862037e+05  2.872664e+03  7.451584e+05   9.663883e+02
min    4.000000e+01  1.000000e+00 -6.000000e+02   0.000000e+00
25%    2.021060e+05  1.160000e+02  0.000000e+00   1.430000e+03
50%    3.856900e+05  3.300000e+02  1.600000e+05   1.920000e+03
75%    7.227500e+05  1.040000e+03  3.450000e+05   2.540000e+03
max    9.906000e+05  9.689000e+03  3.000000e+07   4.816000e+04

In [28]:
#pd.plotting.scatter_matrix(sales_data)    # takes a long time, returned array([<plt object1>,...])   but it also cleared out the memory of earlier cells
sd = sd.loc[sd['SqFtTotLiving']>0]  #already run successfully:  no empty lots
sd = sd.loc[sd['SalePrice'] > 0]
sd['PricePerSqFt'] = sd['SalePrice']/sd['SqFtTotLiving']
sd = sd.loc[sd['PricePerSqFt'] > 10]
#save_file = sd

In [29]:
sd.info()
sd.shape


<class 'pandas.core.frame.DataFrame'>
Int64Index: 860123 entries, 0 to 1437520
Data columns (total 7 columns):
Major            860123 non-null float64
Minor            860123 non-null float64
DocumentDate     860123 non-null object
SalePrice        860123 non-null int64
SqFtTotLiving    860123 non-null int64
ZipCode          860123 non-null object
PricePerSqFt     860123 non-null float64
dtypes: float64(3), int64(2), object(2)
memory usage: 52.5+ MB


(860123, 7)

In [31]:
sd.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0   138860.0  110.0   08/21/2014     245000           1490   98002   
1   138860.0  110.0   06/12/1989     109300           1490   98002   
6   423943.0   50.0   07/15/1999      96000            960   98092   
7   423943.0   50.0   01/08/2001     127500            960   98092   
11  403700.0  715.0   07/03/2013     464500           1780   98008   

    PricePerSqFt  
0     164.429530  
1      73.355705  
6     100.000000  
7     132.812500  
11    260.955056

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [36]:
sd.head()
#sd.info()
sd['ZipCode'].value_counts()
#sd.loc[sd[len(sd['ZipCode']) != 5], : ]

98042         25570
98038         21599
98115         21189
98023         20342
98006         20095
98058         19571
98052         19301
98117         19090
98103         17736
98118         17472
98034         17422
98133         16572
98074         16192
98033         15800
98155         15384
98059         14925
98056         14452
98031         13764
98092         13467
98125         13384
98001         13199
98053         13018
98003         12718
98075         12214
98168         11935
98178         11780
98106         11743
98008         11638
98029         11207
98027         11005
              ...  
98006-3954        1
98176             1
98052-1963        1
98075-6010        1
98028-4712        1
98059-7428        1
95055             1
98074-4092        1
98045-9504        1
98075-9565        1
98028-4505        1
98059-7120        1
8126              1
98054             1
98031-3173        1
98013.0           1
98044             1
98028-8908        1
98075-9517        1


In [40]:
sd['ZipCode'].head()

0     98002
1     98002
6     98092
7     98092
11    98008
Name: ZipCode, dtype: object

In [58]:
int("5432.")

ValueError: invalid literal for int() with base 10: '5432.'

In [57]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def become_integer(x):
    try:
        temp = x[:5]
        return temp
    except ValueError:
        return None
    

temp = sd['ZipCode'].apply(become_integer)
#sales_data.loc[np.isnan(sales_data['ZipCode']) == True, 'ZipCode']
#sales_data['ZipCode']
#temp.loc[temp['ZipCode'] == None]
#temp.astype(int)
#temp.loc[len(temp[) != 5]
#sd = temp.loc[temp.notnull() != False]
temp.head()

0     98002
1     98002
6     98092
7     98092
11    98008
Name: ZipCode, dtype: object

In [ ]:
# formatting
#   = sales_data.loc[~sales_data['ZipCode'].isna(), :]

In [60]:
sd['ZipCode'] = temp
sd.loc[~sd['ZipCode'].isna(), :]

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0        138860.0   110.0   08/21/2014     245000           1490   98002   
1        138860.0   110.0   06/12/1989     109300           1490   98002   
6        423943.0    50.0   07/15/1999      96000            960   98092   
7        423943.0    50.0   01/08/2001     127500            960   98092   
11       403700.0   715.0   07/03/2013     464500           1780   98008   
12       403700.0   715.0   02/21/2013     357000           1780   98008   
13       403700.0   715.0   10/13/1995     142000           1780   98008   
14       403700.0   715.0   02/22/2007     528000           1780   98008   
16        98400.0   380.0   03/23/1994     139950           2050   98058   
17        98400.0   380.0   03/29/2017     439500           2050   98058   
19       330395.0  1610.0   06/07/2012     335000           2210   98038   
20       330395.0  1610.0   08/26/2013     369900           2210   98038   
21       330395.0  1610.0   06/27/1994     192950           2210   98038   
22       330395.0  1610.0   08/22/2013     369900           2210   98038   
23       330395.0  1610.0   11/10/1993     107500           2210   98038   
25       675260.0   940.0   03/06/1998     209950           2040   98031   
26       675260.0   940.0   08/01/1995     191950           2040   98031   
27       675260.0   940.0   09/13/2004     289950           2040   98031   
29       795508.0   170.0   04/09/2001     174000            980   98058   
31       795508.0   170.0   10/31/1996     135000            980   98058   
32       795508.0   170.0   04/30/2004     210000            980   98058   
33       795508.0   170.0   12/21/2012     148000            980   98058   
34       638580.0   110.0   07/14/2015     190000           2130   98188   
36       638580.0   110.0   12/21/2015     338000           2130   98188   
38       252206.0  9011.0   02/21/1997     300000           2530   98051   
39       252206.0  9011.0   11/28/2000     390000           2530   98051   
40       321165.0    50.0   05/10/2004    2100000           1800   98001   
42       321165.0    50.0   10/24/2013     215000           1800   98001   
44       321165.0    50.0   09/20/2004     270000           1800   98001   
45       321165.0    20.0   05/10/2004    2100000           1800   98001   
...           ...     ...          ...        ...            ...     ...   
1437478  928170.0     5.0   12/11/1986      89950           1630   98116   
1437479  783280.0    60.0   08/21/1986      67500           1200   98146   
1437481  809260.0    80.0   08/20/2001     145000           1170   98042   
1437484  344800.0  1500.0   07/12/1996     174000            680   98125   
1437485  177421.0   370.0   06/24/1993     233000           2930   98077   
1437487  278731.0   900.0   10/18/1982      78000           1590   98031   
1437488   12605.0  9008.0   10/01/1993     110000            790   98072   
1437489   42506.0  9158.0   03/07/1988      35000           1920   98053   
1437490  955020.0  4120.0   03/24/1994     156500           1370   98105   
1437491  220750.0   450.0   05/19/1993     199000           1430   98102   
1437492  546878.0   130.0   03/23/2005     309900           1940   98042   
1437493  251700.0   900.0   07/05/2005     272195           1920   98042   
1437498  881990.0   515.0   09/25/1985      63500            870   98105   
1437499  222505.0  9207.0   10/09/1995     246500           2570   98005   
1437500  222505.0  9207.0   10/09/1995     246500           1070   98005   
1437502  541860.0    20.0   07/23/2002     450000           2890   98115   
1437504  172606.0  9181.0   07/26/1983      87950           2320   98077   
1437505  293500.0   100.0   01/23/1987      66850           1190   98001   
1437508   80900.0  1440.0   10/19/1988     147000           1730   98109   
1437509  247338.0   310.0   07/27/1987      83000           1870   98058   
1437510   11410.0   629.0   01/26/1998     171000           

### 3. Add a column for PricePerSqFt



In [ ]:
#done above

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [62]:
sd.loc[sd['DocumentDate'].str.contains('2019')]

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
605      214090.0   110.0   03/06/2019     390000           1780   98042   
651      367890.0    30.0   03/13/2019     923000           2550   98144   
868      350160.0   125.0   02/06/2019     935000           2460   98117   
1063     882790.0   870.0   02/15/2019     890000           2380   98105   
1341      30200.0   645.0   03/04/2019     380000           1750   98001   
1657      82007.0  9027.0   02/15/2019     895000           3160   98022   
1997     352106.0  9028.0   03/21/2019     595000           3580   98022   
2555      98400.0   450.0   02/20/2019     409950           1850   98058   
2821     945920.0   125.0   01/29/2019    1750000           1400   98118   
3541     152104.0  9136.0   01/24/2019     268000           1551   98001   
4362      70570.0   490.0   03/20/2019     470000           1670   98038   
5347     869136.0   290.0   02/13/2019    1237000           4770   98075   
5359     186240.0   105.0   03/19/2019     380000            800   98117   
5468     808335.0    40.0   01/09/2019     369950           1660   98055   
5724     953010.0   160.0   03/05/2019     950000           1490   98107   
6058     807903.0   400.0   02/14/2019     650000           2560   98059   
6322     390508.0  1800.0   03/04/2019     800000           2170   98029   
7785     880490.0   930.0   01/02/2019    1875000           2490   98122   
8222     144680.0  1055.0   03/20/2019     330000            820   98168   
8279     723150.0  1645.0   03/08/2019     407000            970   98055   
8586     334570.0   212.0   03/13/2019     795000           2340   98056   
8609     327480.0   560.0   01/18/2019     108560            890   98144   
8743     247520.0   700.0   01/29/2019     445000           1760   98055   
9821     302505.0  9027.0   02/21/2019    5165000           5320   98004   
11434    545830.0   175.0   03/18/2019    2850000           1150   98109   
12242    420240.0   245.0   01/07/2019     307000           1480   98055   
13373    383206.0  1090.0   02/18/2019     390000           1940   98042   
13786    752510.0   490.0   03/18/2019     285000           1350   98052   
14690     82006.0  9038.0   02/13/2019     825000           2190   98022   
14691     82006.0  9038.0   02/13/2019     825000           1304   98022   
...           ...     ...          ...        ...            ...     ...   
1422627  262206.0  9028.0   02/07/2019     500000           1640   98051   
1422628  220720.0    55.0   03/06/2019    1690000           1030   98007   
1423096  690970.0   290.0   01/03/2019    1041100            660   98144   
1423097  690970.0   290.0   01/03/2019    1041100           2050   98144   
1424257  247300.0  2030.0   03/13/2019     537000           2270   98058   
1424661   20390.0   330.0   02/27/2019     450000           1630   98053   
1424839  660006.0    20.0   03/25/2019     626400           2330   98146   
1424896  432120.0   795.0   03/11/2019     749900            840   98126   
1424897  935090.0   180.0   03/27/2019    1390000           2360   98040   
1425175  277910.0    20.0   01/03/2019     545000           1650   98122   
1425201  123630.0   267.0   02/09/2019     670000            940   98033   
1425728  132205.0  9037.0   02/28/2019     590000           1050   98042   
1425729  132205.0  9037.0   02/28/2019     590000           2070   98042   
1428008  293360.0    20.0   03/05/2019     325000           1530   98166   
1428595  660006.0    10.0   01/11/2019     575000           2330   98146   
1428788  547010.0    30.0   03/13/2019     420000           2620   98042   
1430076  889900.0   410.0   01/16/2019     449950           2140   98055   
1430494  404840.0   365.0   03/21/2019     594000           2160   98059   
1430586   47500.0   715.0   01/23/2019    1050000           3180   98107   
1430742  102306.0  9047.0   02/04/2019     749000           2090   98027   
1432030  277660.0   116.0   03/18/2019     740000           

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [72]:
seattle_zips = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98121, 
                98122, 98125, 98126, 98133, 98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]
seattle_home_sales = sd[sd['ZipCode'].isin([str(x) for x in seattle_zips])]

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [78]:
final = seattle_home_sales.loc[seattle_home_sales['DocumentDate'].str.contains('2019')]
# it is 507.872
final['PricePerSqFt'].mean()

507.8724846169785